File: core/src/main/java/y2k/joyreactor/presenters/ProfilePresenter.java
package y2k.joyreactor.presenters;

import y2k.joyreactor.platform.Navigation;
import y2k.joyreactor.Profile;
import y2k.joyreactor.services.ProfileService;

/**
 * Created by y2k on 9/30/15.
 */
public class ProfilePresenter {

    private ProfileService service;

    public ProfilePresenter(View view, ProfileService service) {
        this.service = service;

        view.setBusy(true);
        service.get()
                .subscribe(profile -> {
                    view.setProfile(profile);
                    view.setBusy(false);
                }, e -> {
                    e.printStackTrace();
                    Navigation.getInstance().switchProfileToLogin();
                });
    }

    public void logout() {
        service.logout().subscribe(_void -> Navigation.getInstance().switchProfileToLogin());
    }

    public interface View {

        void setProfile(Profile profile);

        void setBusy(boolean isBusy);
    }
}

File: core/src/main/kotlin/y2k/joyreactor/presenters/ProfilePresenter.kt
package y2k.joyreactor.presenters

import y2k.joyreactor.model.Profile
import y2k.joyreactor.common.subscribeOnMain
import y2k.joyreactor.platform.Navigation
import y2k.joyreactor.services.ProfileService

/**
 * Created by y2k on 9/30/15.
 */
class ProfilePresenter(
    private val view: ProfilePresenter.View,
    private val service: ProfileService) {

    init {
        view.setBusy(true)
        service
            .getProfile()
            .subscribeOnMain({
                view.setProfile(it)
                view.setBusy(false)
            }, {
                it.printStackTrace()
                Navigation.instance.switchProfileToLogin()
            })
    }

    fun logout() {
        service.logout().subscribeOnMain { Navigation.instance.switchProfileToLogin() }
    }

    interface View {

        fun setProfile(profile: Profile)

        fun setBusy(isBusy: Boolean)
    }
}